x-pack/legacy/plugins/siem/server/lib/timeline/routes/export_timelines_route.ts

```diff
@@ -4,34 +4,29 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+import { schema } from '@kbn/config-schema';
+
 import { set as _set } from 'lodash/fp';
 import { IRouter } from '../../../../../../../../src/core/server';
 import { LegacyServices } from '../../../types';
-import { ExportTimelineRequestParams } from '../types';
 
-import {
-  transformError,
-  buildRouteValidation,
-  buildSiemResponse,
-} from '../../detection_engine/routes/utils';
+import { transformError, buildSiemResponse } from '../../detection_engine/routes/utils';
 import { TIMELINE_EXPORT_URL } from '../../../../common/constants';
 
-import {
-  exportTimelinesSchema,
-  exportTimelinesQuerySchema,
-} from './schemas/export_timelines_schema';
-
 import { getExportTimelineByObjectIds } from './utils/export_timelines';
 
 export const exportTimelinesRoute = (router: IRouter, config: LegacyServices['config']) => {
   router.post(
     {
       path: TIMELINE_EXPORT_URL,
       validate: {
-        query: buildRouteValidation<ExportTimelineRequestParams['query']>(
```

Comment:
@angorayc I was imagining that once Joi was gone, this `buildRouteValidation` function would be modified to take an io-ts schema and perform whatever decoding/validation needs to happen.

It looks like we're currently loosely typing the request with `@kbn/config-schema` in the validation step, and then refining that further within the route handler with `io-ts`; is it possible to simply use `io-ts` in the validation step, similar to what `buildRouteValidation` was doing? If we can we should, as that removes the possibility of interacting with these "loose" types in the handler and losing type safety there (and removes the double validation).

Reply:
@rylnd , @angorayc Here is my draft PR where I am doing what @rylnd is suggesting for detection engine:

Example of where I validate incoming using io-ts to search for:

```ts
validate: {
        body: buildRouteValidationIoTS<CreateListsSchema>(createListsSchema),
      },
```

My function which performs an io-ts validation similar to the one above it that @rylnd is using:
https://github.com/elastic/kibana/pull/62552/files#diff-8b8c8b8c0404c58c669ab9551d6f2e8bL238-L239

This is very draft so again, refactoring/changing things afterwards is good but if it helps I can begin pulling in some of these utilities as soon as this the beginning of this week @rylnd to start helping reducing our usage of schemas and establishing better patterns.

It might be best I begin pulling in my pieces from draft mode in smaller PR's Monday and getting those reviewed. Then we can migrate any other things like this after 7.7.0 is released as the time is running out quickly?

I don't want to hold anything up though. Refactoring can happen after this PR too and my pieces.

Reply:
That's beautiful @FrankHassanabad, @rylnd  I've been trying to figure this out for a while! If I'm going to reuse it in timeline's route once it's merged, would you recommend importing from this file or there's better way of how we could reuse it? As currently I use lots of stuff from detection engine etc., not sure if we could have an utils for all the routes? Or is it possible if we implement this part in middleware so we can just provide the validation schemas in each route without implementing `buildRouteValidationIoTS `?

Reply:
Yeah it's tricky at the moment but we should be promoting things "up" as you are suggesting and out of detection engine as they see more re-use. In your PR's you can move things around as you see fit or if we have some duplication we can deconflict as well.

It's tricky as we are all working in the same areas but getting code in first and working is probably easier and then refactors second once people move to other areas maybe?

I am fine with either approach and I like your ideas. The middle ware idea is a good one but I haven't written those before with Hapi or the Kibana platform but am open to any and all ideas if you want to give it a try.

Reply:
Thanks @FrankHassanabad , at the same time I found some code useful for validation,
https://elastic.slack.com/archives/CCJQY1NS0/p1584130424322000
It matches what I thought and fixes the incorrect error message I had, what do you think?

Reply:
@angorayc yep, that code looks very similar to what Frank spiked out and what I was imagining, so it sounds like we're all on the same page. I'm less familiar with io-ts but I'd be happy to pair whenever you pick this back up!

Reply:
@rylnd, @FrankHassanabad  cool, I've updated my validation with the way as we discussed, and also lifted some common functions up from detection engines. Please have another look when you have time, thanks.

Reply:
@angorayc this is looking better, but we shouldn't need to provide both a type and a schema to our validation factory: one of the benefits of switching to io-ts is that we can generate types from the schema. I've updated @FrankHassanabad's example to require only the schema:

```ts
export const buildRouteValidationIoTS = <T extends t.Mixed, A = t.TypeOf<T>>(
  schema: T
): RouteValidationFunction<A> => (payload: object, { ok, badRequest }) => {
  const decoded = schema.decode(payload);
  const checked = exactCheck(payload, decoded);
  const left = (errors: t.Errors): RouteValidationFunctionReturn<A> => {
    return badRequest(formatErrors(errors).join(','));
  };
  const right = (output: A): RouteValidationFunctionReturn<A> => {
    return ok(output);
  };
  return pipe(checked, fold(left, right));
};
```

Reply:
Note: if you need to override the output type above (`A`), you'll have to do something like:

```ts
buildRouteValidation<t.Any, OverridingResponseType>(schema)
```

It feels like there should be a way to support overriding the type with a single generic argument, but I hit the limits of my typescript knowledge trying to make that work.

Reply:
Thanks @rylnd , I've applied the suggested type and it works well. I agree that we should warning user their if their input got dropped, and given that @FrankHassanabad has implemented the solution nicely, I'd like to park here and creating an issue to keep track of this.
https://github.com/elastic/siem-team/issues/615

Reply:
I found in x-pack/plugins/case/common/api/runtime_types.ts Line. 37
Seems it's doing similar thing as exact check, which means when validation error occurs, it return the original object.

e.g.:
I updated the buildValidation:

```
export const buildRouteValidation = <T extends rt.Mixed, A = rt.TypeOf<T>>(
  schema: T
): RouteValidationFunction<A> => (
  inputValue: unknown,
  validationResult: RouteValidationResultFactory
) =>
  pipe(
    excess(schema).decode(inputValue),
    fold<rt.Errors, A, RequestValidationResult<A>>(
      (errors: rt.Errors) => validationResult.badRequest(failure(errors).join('\n')),
      (validatedInput: A) => validationResult.ok(validatedInput)
    )
  );
```

Use case:
I sent this as request body: 
```{ id: 'someId' }```
but it's expecting
```{ id: string[] }```


`Before`
Expected: "Invalid value undefined supplied to : { ids: Array<string> }/ids: Array<string>"
`After`
Received: "Invalid value {\"id\":\"someId\"} supplied to : { ids: Array<string> }, excess properties: [\"id\"]"

Reply:
Ok, we can address this as part of #63525.